src/shared/src/types.rs

```diff
@@ -100,3 +100,57 @@ pub mod custom_token {
         Icrc(LedgerId),
     }
 }
+
+/// Types specifics to the user profile.
+pub mod user_profile {
+    use candid::{CandidType, Deserialize, Principal};
+    use std::collections::BTreeMap;
+
+    pub type CredentialType = String;
+
+    #[derive(CandidType, Deserialize, Clone, Eq, PartialEq, Debug)]
+    pub struct UserCredential {
+        pub credential_type: CredentialType,
+        pub verified_date_timestamp: Option<u64>,
+        pub expire_date_timestamp: Option<u64>,
+    }
+
+    // Used in the endpoint
+    #[derive(CandidType, Deserialize, Clone, Eq, PartialEq, Debug)]
+    pub struct UserProfile {
```

Comment:
The user profile should, in my opinion, similarly to `Token` and `CustomToken` contains a `version: Option<Version>` and implement and the `TokenVersion` should be implemented for it.

Not sure if we also want to introduce a version field within the `UserCredential`

Reply:
Oh, that adds quite come complexity. How does that work? What's the advantage of having this?

Reply:
> What's the advantage of having this?

Prevent data overwrite if the same entity is edited multiple times at the same times.

Think at a user who use the dApp at the same time on their mobile phone and laptop for example.

> How does that work?

Basically the caller must provide the version that matches the entity in the smart contract otherwise the update is rejected.

Kind of similar to what's done in SQL database.

Reply:
Thanks! I added the `version` field.